Update Release notes script to new Layout with Jira Epics

SaikrishnaBairamoni opened the pull request:
<!-- Thanks for the contribution, this is awesome. -->

# PR Details
This PR improves the release notes script by improving the layout to include Jira epics and PR details, following the template outlined here: https://usdot-carma.atlassian.net/wiki/spaces/CRMPLT/pages/2955804678/Release+Notes+-+Output+File+Template. It also adds detailed docstrings for clarity and addresses Pylint and SonarLint warnings.

Example Template
```md
## carma-msgs - Test1.1

**List of Jira Epics**
* ARC-73: STOL organization-wide DevOps processes. (Status: To Do): This epic is meant to capture improvements/maintenance of our STOL organization wide DevOps processes. This includes managing repository settings and branch protection rules and maintaining generic CI/CD processes used by projects. 
NOTE: Project Specific Devops tasks should be track under project scope. Examples of this include tool repository CI/CD processes and project releases. (GitHub PRs #239, #237)

**List of other Jira Items (e.g. Story/Bug/Anomaly/Task)**
* j2735_v2x_msgs package is not forward compatible with Humble (Jira Story : CF-813) - Epic missing

**List of GitHub Issues**
No GitHub issues found

**List of GitHub PRs (no Jira Item or GitHub Issue)**
* Sync develop to master (Pull Request [#232](https://github.com/usdot-fhwa-stol/carma-msgs/pull/232))
* Fix CI/Build to remove branch coupling (Pull Request [#233](https://github.com/usdot-fhwa-stol/carma-msgs/pull/233))

**List of Orphaned Commits**
* Commit: Sync develop to master (#232) (Commit [9991dd])
* Commit: Fix GitHub action triggers (Commit [8b6fde])
* Commit: Updates (Commit [171d54])
* Commit: Fix target directory for cloned ros1_bridge repository to ensure it is built in final image (#237) (Commit [7bb468])
* Commit: Update checkout script to account for carma-master and carma-develop branches (Commit [bacf59])
* Commit: Refactor static member variable access to resolve compiler  (#227) (Commit [81d861])
* Commit: update scripts to pass arguments and build with build-image.sh (#239) (Commit [2c6bb0])
* Commit: Fix CI/Build to remove branch coupling (#233) (Commit [ea1aa7])
* Commit: Add branch argument to ci checkout.sh call (Commit [e86158])
* Commit: Update github actions and set default value for BRANCH in checkout script (Commit [9d4d16])
* Commit: Correct bash syntax error (Commit [751eef])
```
<!--- Describe your changes in detail -->

## Related GitHub Issue

<!--- This project only accepts pull requests related to open GitHub issues or Jira Keys -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please DO NOT name partially fixed issues, instead open an issue specific to this fix -->
<!--- Please link to the issue here: -->

## Related Jira Key
[ARC-206](https://usdot-carma.atlassian.net/browse/ARC-206)
<!-- e.g. CAR-123 -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.


[ARC-206]: https://usdot-carma.atlassian.net/browse/ARC-206?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

SaikrishnaBairamoni committed:
Update script to add docstrings and format output

dan-du-car commented on generate_release_notes.py:
```diff
@@ -80,33 +98,38 @@ def get_parent_epic(jira_issue, jira_url, jira_email, jira_token):
+
+    Args:
+        github_repo (Repository): The GitHub repository object.
+        pr_number (int): The number of the pull request.
```
This is the number of pull request or id of a pull request?

SaikrishnaBairamoni replied:
Pr_number arg is related to Pull request number we are using the PR number and  to capture the PR details, I will update the docstring

dan-du-car commented on generate_release_notes.py:
```diff
@@ -80,33 +98,38 @@ def get_parent_epic(jira_issue, jira_url, jira_email, jira_token):
     if issue_body:
-        # Get Jira keys only from the "Related Jira Key" section of PR's
-        jira_key_match = re.findall(r'Related Jira Key.*?(\[.*?\])?\(?([A-Z]+-\d+)\)?', issue_body, re.DOTALL)
+        jira_key_match = re.findall(r'Related Jira Key.*?(\[[^\]]*\])?\(?([A-Z]+-\d+)\)?', issue_body, re.DOTALL)
```
Could you add a comments with sample JIRA key? Why this regex update?

SaikrishnaBairamoni replied:
I added example in docstring and I updated the regex to resolve Pylint/SonarLint warnings and only capture with in squarebrackets

dan-du-car commented on generate_release_notes.py:
```diff
@@ -80,33 +98,38 @@ def get_parent_epic(jira_issue, jira_url, jira_email, jira_token):
-        # If no Jira keys found, check for GitHub Issues from the "Related GitHub Issue" section
         if not jira_keys:
-            github_issue_match = re.findall(r'Related GitHub Issue.*?(\[.*?\])?\(?#(\d+)\)?', issue_body, re.DOTALL)
+            github_issue_match = re.findall(r'Related GitHub Issue.*?(\[[^\]]*\])?\(?#(\d+)\)?', issue_body, re.DOTALL)
```
The same comments as above

SaikrishnaBairamoni replied:
I added example in docstring and I updated the regex to resolve Pylint/SonarLint warnings and only capture with in squarebrackets

dan-du-car commented on generate_release_notes.py:
```diff
@@ -125,173 +148,241 @@ def get_repo_list(github_org, github):
+    """
+    notes_content = f"\n\n## {name} - {version}\n"
+
+    # List of Jira Epics
```
Are these epics here referring to JIRA stories?

SaikrishnaBairamoni replied:
These are referring to Jira Epics we are capturing epics by using Jira stories if epic is not found we will just keep Jira story under list of Other Jira Items. Our Main goal is to capture Epics information.

dan-du-car commented on generate_release_notes.py:
```diff
@@ -125,173 +148,241 @@ def get_repo_list(github_org, github):
     if issue_titles_other:
-        release_notes += "\n\n#### Other Issues\n"
-        release_notes += "* " + "\n* ".join(sorted(set(issue_titles_other)))
+        notes_content += "\n### List of Other Jira Items (Bugs, Tasks, Anomalies)\n"
```
The tempate for this title:
```
List of other Jira Items (e.g. Story/Bug/Anomaly/Task) (where no parent Jira Epic assigned)
```

dan-du-car replied:
Need to be bold.

SaikrishnaBairamoni replied:
Updated

dan-du-car commented on generate_release_notes.py:
```diff
@@ -125,173 +148,241 @@ def get_repo_list(github_org, github):
+            epic_description = epic_parts[2] if len(epic_parts) > 2 else ""
+            pr_number = pr_mapping.get(epic_key, "N/A")
+            if epic_description:
+                notes_content += f"Epic ### {epic_key}: {epic_title}: {epic_description} (GitHub PR #{pr_number})\n"
```
Maybe remove this `###`
Do we need bullet point before `Epic` with aterisk `*`?

dan-du-car replied:
Title and key need to be bold.

SaikrishnaBairamoni replied:
Updated it

dan-du-car commented on generate_release_notes.py:
```diff
@@ -125,173 +148,241 @@ def get_repo_list(github_org, github):
     if issue_titles_epics:
-        release_notes += "\n\n#### Jira Epics\n"
-        release_notes += "* " + "\n* ".join(sorted(set(issue_titles_epics)))
+        notes_content += "\n### List of Jira Epics\n"
```
Make this title bold with `**`

SaikrishnaBairamoni replied:
updated

dan-du-car commented on generate_release_notes.py:
```diff
@@ -125,173 +148,241 @@ def get_repo_list(github_org, github):
+        notes_content += "\n### No Other Jira Items Found\n"
+
+    if pull_requests_missing_epics:
+        notes_content += "\n### List of GitHub PRs Missing Epics and Issues\n"
```
Maybe use JIRA Item to be more specific rather than Epic? Add GitHub before Issues to be clear.

dan-du-car replied:
Bold

SaikrishnaBairamoni replied:
Updated

dan-du-car commented on generate_release_notes.py:
```diff
@@ -125,173 +148,241 @@ def get_repo_list(github_org, github):
+            else:
+                pr_title = pr
+                pr_description = "No description"
+            notes_content += f"PR ###: {pr_title}: {pr_description} (Commit ###)\n"
```
remove ###

SaikrishnaBairamoni replied:
Updated

dan-du-car commented on generate_release_notes.py:
```diff
@@ -125,173 +148,241 @@ def get_repo_list(github_org, github):
     if commit_only:
-        release_notes += "\n\n#### Commits Missing Issues\n"
-        release_notes += "* " + "\n* ".join(sorted(commit_only))
+        notes_content += "\n### List of Orphaned Commits\n"
```
Bold title

SaikrishnaBairamoni replied:
Updated

dan-du-car commented on generate_release_notes.py:
```diff
@@ -125,173 +148,241 @@ def get_repo_list(github_org, github):
+        notes_content += "\n### List of Orphaned Commits\n"
+        for commit in commit_only:
+            commit_title = commit.split(': ', 1)[0]
+            notes_content += f"Commit ###: {commit_title}\n"
```
Title is bold and need to add description if exist.

SaikrishnaBairamoni replied:
Updated

SaikrishnaBairamoni committed:
address comments

dan-du-car commented:
- The "List of Jira Epics" title is smaller than content starting with ARC-73. Also the content description starting from "This epic is meant to" should not be bold
- Can we indent the content  bullet point to highlight that it is a section of "List of Jira Epics"? The same intent for all bullet points.
- "List of GitHub PRs Missing Jira Epics and GitHub issues":  Do we need "commit ###" if the we have PR for it? The PR number is missing.
![image](https://github.com/user-attachments/assets/0e2eb7de-6ca0-42ce-a54c-ae8f8a1dd654)

Indentation is incorrect, see below:
![image](https://github.com/user-attachments/assets/e83ffe1b-9be7-45bb-aab7-4d1095135dc0)

paulbourelly999 commented on generate_release_notes.py:
```diff
@@ -125,173 +149,241 @@ def get_repo_list(github_org, github):
 if __name__ == "__main__":
     parser = argparse.ArgumentParser()
-    parser.add_argument("--github-token", required=True)  
+    parser.add_argument("--github-token", required=True)
```
Please add descriptions for these arguments and default values where it makes sense. Like stable branch or organizations or jira URL since those will likely always be the same.

SaikrishnaBairamoni replied:
Added default values and descriptions

SaikrishnaBairamoni committed:
Assress comments round 2

SaikrishnaBairamoni committed:
Address Comment by adding parser default and descriptions.

paulbourelly999 commented on generate_release_notes.py:
```diff
@@ -80,33 +98,39 @@ def get_parent_epic(jira_issue, jira_url, jira_email, jira_token):
-        # If no Jira keys found, check for GitHub Issues from the "Related GitHub Issue" section
         if not jira_keys:
-            github_issue_match = re.findall(r'Related GitHub Issue.*?(\[.*?\])?\(?#(\d+)\)?', issue_body, re.DOTALL)
+            github_issue_match = re.findall(r'Related GitHub Issue.*?(\[[^\]]*\])?\(?#(\d+)\)?', issue_body, re.DOTALL)
```
Why is this regex different from the one above?

SaikrishnaBairamoni replied:
I changed this regex accordinglg from pylint warnings as takes only characters inside the square brackets which are matched but it extracts GitHub issue numbers from the pull request body with no effect.

paulbourelly999 commented on generate_release_notes.py:
```diff
@@ -125,180 +149,258 @@ def get_repo_list(github_org, github):
-        logging.error(e)
-        logging.error(msg_failure)
+    except GithubException as error:
+        logging.error("%s", error)
```
Please combine into single error log

SaikrishnaBairamoni replied:
Updated into single error log statement.

paulbourelly999 commented on generate_release_notes.py:
```diff
@@ -125,180 +149,258 @@ def get_repo_list(github_org, github):
+def get_release_notes(name, version, issue_titles_epics,
+                      issue_titles_other, github_issues, pr_missing_jira_issues, commit_only, pr_mapping):
+    """
+    Format release notes for the given repository.
```
Generate formatted release notes for given repository.

SaikrishnaBairamoni replied:
Updated

paulbourelly999 commented on generate_release_notes.py:
```diff
@@ -125,180 +149,258 @@ def get_repo_list(github_org, github):
                     else:
-                        commit_url = found_commit.commit.html_url[:-34]
-                        commit_title = "{} (Commit [{}])".format(found_commit.commit.message.strip().split('\n', 1)[0], found_commit.commit.sha[:6])
+                        commit_title = "{} (Commit [{}])".format(
```
Why do we need the individual commit titles. I assumed we would only need Jira or PR titles.

SaikrishnaBairamoni replied:
We are included commit titles in the orphaned commit detials section to capture changes made to the branch that aren't associated to a PR or Jira issue.

paulbourelly999 commented on generate_release_notes.py:
```diff
@@ -125,180 +149,258 @@ def get_repo_list(github_org, github):
+        issue_titles_other (list): Titles of other issues.
+        github_issues (list): GitHub issues.
+        pr_missing_jira_issues (list): PRs missing Jira items or GitHub issues.
+        commit_only (list): Commits missing issues.
```
What is a commit missing an issue? Is that a PR missing an issue? If we merge a PR without an issue and the PR includes multiple commits, will we get a bunch of commits missing issues or a PR missing Jira/Github issue or both?

SaikrishnaBairamoni replied:
So commit missing issue refers to a commit that was made directly to the branch and not associated with any pull request or linked with Jira/GitHub issues. So If no PR is found for a commit, the commit is added to the commit_only list and these are added under orphaned commits at the end. PR missing an issue refers to when a PR is found but does not have any linked Jira/GitHub issues. we will have some commits missing with issues that are captured under each repo but I think as this will be in our intermediate output file and we just use this to make a final notes.

paulbourelly999 commented on generate_release_notes.py:
```diff
@@ -125,180 +149,258 @@ def get_repo_list(github_org, github):
+        name (str): The repository name.
+        version (str): The release version.
+        issue_titles_epics (set): Titles of unique Jira epics.
+        issue_titles_other (list): Titles of other issues.
```
Is this other Jira Issues?

SaikrishnaBairamoni replied:
Yes, Updated

paulbourelly999 commented on generate_release_notes.py:
```diff
@@ -125,180 +149,258 @@ def get_repo_list(github_org, github):
+            epic_description = epic_parts[2] if len(epic_parts) > 2 else ""
+            pr_number = pr_mapping.get(epic_key, "N/A")
+            if epic_description:
+                notes_content += f"* {epic_key}: {epic_title}: {epic_description} (GitHub PR #{pr_number})\n"
```
An epic may include multiple PRs. Does the map account for that?

SaikrishnaBairamoni replied:
Updated script to capture multiple PRs of epic

paulbourelly999 commented on generate_release_notes.py:
```diff
@@ -125,180 +149,258 @@ def get_repo_list(github_org, github):
                                         if epic_title:
-                                            issue_titles_epics.append(f"{epic_key} - {epic_title}: {epic_description}")
+                                            issue_titles_epics.add(f"{epic_key} - {epic_title}: {epic_description}")
+                                            pr_mapping[epic_key] = pr.number
```
This does not account for multiple PRs making up  a single epic. As currently written each epic will only be associated with a single PR number and it will be the last PR for that epic. This code needs to be updates to have the `pr_mapping` be a dictionary of `epic_keys` to array of pr numbers

SaikrishnaBairamoni replied:
Updated

SaikrishnaBairamoni committed:
Address comments

paulbourelly999 committed:
Fix script bugs
+ PRs with Jira Issues without epics will be added to List of other Jira Items
+ Removing duplicate information for epic description and status
+ Using query parameters to reduce fields returned by Jira Issue API calls

paulbourelly999 committed:
Updates

SaikrishnaBairamoni committed:
fix commit list to flag only direct commits as orphaned

SaikrishnaBairamoni commented:
@paulbourelly999 I pushed change added pr_list.totalCount > 0 to check if a commit is part of a PR, and moved the else block inside the if to make sure it only runs when there are no associated PRs. If there’s an error it will log the commit under skipped PRs as usual. I tested it out across release/omega, develop against master and working fine. Below is the sample output generated. Thanks!
[Release_Notes_directcommits.md.zip](https://github.com/user-attachments/files/17298413/Release_Notes_directcommits.md.zip)
